pkg/storage/s3.go

```diff
@@ -338,44 +383,68 @@ func (s *S3Storage) UploadProviderReleaseFiles(ctx context.Context, namespace, n
 		return fmt.Errorf("name argument is empty")
 	}
 
-	prefix, err := providerStoragePrefix(s.bucketPrefix, internalProviderType, "", namespace, name)
-	if err != nil {
-		return err
-	}
-
+	prefix := providerStoragePrefix(s.bucketPrefix, internalProviderType, "", namespace, name)
 	key := filepath.Join(prefix, filename)
+	return s.upload(ctx, key, file, false)
+}
+
+func (s *S3Storage) signingKeys(ctx context.Context, pt providerType, hostname, namespace string) (*core.SigningKeys, error) {
+	if namespace == "" {
+		return nil, fmt.Errorf("namespace argument is empty")
+	}
+	key := signingKeysPath(s.bucketPrefix, pt, hostname, namespace)
 	exists, err := s.objectExists(ctx, key)
 	if err != nil {
-		return err
-	} else if exists {
-		return ErrProviderAlreadyExists
-	}
-
-	input := &s3.PutObjectInput{
-		Bucket: aws.String(s.bucket),
-		Key:    aws.String(key),
-		Body:   file,
+		return nil, err
+	} else if !exists {
+		return nil, core.ErrObjectNotFound
 	}
 
-	if _, err = s.uploader.Upload(ctx, input); err != nil {
-		return fmt.Errorf("failed to upload provider: %w", err)
+	signingKeysRaw, err := s.download(ctx, key)
+	if err != nil {
+		return nil, fmt.Errorf("failed to download signing_keys.json for namespace %s: %w", namespace, err)
 	}
 
-	return nil
+	return unmarshalSigningKeys(signingKeysRaw)
 }
 
 // SigningKeys downloads the JSON placed in the namespace in S3 and unmarshals it into a core.SigningKeys
 func (s *S3Storage) SigningKeys(ctx context.Context, namespace string) (*core.SigningKeys, error) {
-	if namespace == "" {
-		return nil, fmt.Errorf("namespace argument is empty")
+	return s.signingKeys(ctx, internalProviderType, "", namespace)
```

Comment:
I'm pretty sure the `hostname` must be taken into account here. Otherwise the mirroring wont work. This however implies a signature change of `SigningKeys` as well which impacts GCS and the `uploadProvider` function. Here is a patch:

```
diff --git a/cmd/upload.go b/cmd/upload.go
index fb2bc65..911a850 100644
--- a/cmd/upload.go
+++ b/cmd/upload.go
@@ -150,7 +150,7 @@ func uploadProvider(cmd *cobra.Command, args []string) error {

 	validateCtx, cancelValidateCtx := context.WithTimeout(ctx, 15*time.Second)
 	defer cancelValidateCtx()
-	signingKeys, err := storageBackend.SigningKeys(validateCtx, flagProviderNamespace)
+	signingKeys, err := storageBackend.SigningKeys(validateCtx, "", flagProviderNamespace)
 	if err != nil {
 		return err
 	}
diff --git a/pkg/provider/storage.go b/pkg/provider/storage.go
index 4c79ee1..36f11f4 100644
--- a/pkg/provider/storage.go
+++ b/pkg/provider/storage.go
@@ -19,7 +19,7 @@ type Storage interface {
 	UploadProviderReleaseFiles(ctx context.Context, namespace, name, filename string, file io.Reader) error

 	// SigningKeys downloads and returns the keys for a given namespace from the configured storage backend
-	SigningKeys(ctx context.Context, namespace string) (*core.SigningKeys, error)
+	SigningKeys(ctx context.Context, hostname string, namespace string) (*core.SigningKeys, error)

 	// MigrateProviders is needed for the migration from 0.7.0 to 0.8.0
 	MigrateProviders(ctx context.Context, logger log.Logger, dryRun bool) error
diff --git a/pkg/storage/gcs.go b/pkg/storage/gcs.go
index 5bc117b..f3a5ec0 100644
--- a/pkg/storage/gcs.go
+++ b/pkg/storage/gcs.go
@@ -231,7 +231,7 @@ func (s *GCSStorage) getProvider(ctx context.Context, pt providerType, provider
 	if err != nil {
 		return nil, err
 	}
-	signingKeys, err := s.SigningKeys(ctx, provider.Namespace)
+	signingKeys, err := s.SigningKeys(ctx, provider.Hostname, provider.Namespace)
 	if err != nil {
 		return nil, err
 	}
@@ -370,8 +370,8 @@ func (s *GCSStorage) signingKeys(ctx context.Context, pt providerType, hostname,
 }

 // SigningKeys downloads the JSON placed in the namespace in GCS and unmarshals it into a core.SigningKeys
-func (s *GCSStorage) SigningKeys(ctx context.Context, namespace string) (*core.SigningKeys, error) {
-	return s.signingKeys(ctx, internalProviderType, "", namespace)
+func (s *GCSStorage) SigningKeys(ctx context.Context, hostname string, namespace string) (*core.SigningKeys, error) {
+	return s.signingKeys(ctx, internalProviderType, hostname, namespace)
 }

 func (s *GCSStorage) MirroredSigningKeys(ctx context.Context, hostname, namespace string) (*core.SigningKeys, error) {
diff --git a/pkg/storage/s3.go b/pkg/storage/s3.go
index 5bdbf2f..e1ee105 100644
--- a/pkg/storage/s3.go
+++ b/pkg/storage/s3.go
@@ -281,7 +281,7 @@ func (s *S3Storage) getProvider(ctx context.Context, pt providerType, provider *
 		return nil, err
 	}

-	signingKeys, err := s.SigningKeys(ctx, provider.Namespace)
+	signingKeys, err := s.SigningKeys(ctx, provider.Hostname, provider.Namespace)
 	if err != nil {
 		return nil, err
 	}
@@ -409,8 +409,8 @@ func (s *S3Storage) signingKeys(ctx context.Context, pt providerType, hostname,
 }

 // SigningKeys downloads the JSON placed in the namespace in S3 and unmarshals it into a core.SigningKeys
-func (s *S3Storage) SigningKeys(ctx context.Context, namespace string) (*core.SigningKeys, error) {
-	return s.signingKeys(ctx, internalProviderType, "", namespace)
+func (s *S3Storage) SigningKeys(ctx context.Context, hostname string, namespace string) (*core.SigningKeys, error) {
+	return s.signingKeys(ctx, internalProviderType, hostname, namespace)
 }

 func (s *S3Storage) MirroredSigningKeys(ctx context.Context, hostname, namespace string) (*core.SigningKeys, error) {
```

WDYT?

Reply:
Good catch, thanks for the review! There is already `MirroredSigningKeys()`, so let's use that instead :)
https://github.com/TierMobility/boring-registry/blob/8ad6a194b2246c5dccd13c18a9f3a136b792f8a2/pkg/mirror/storage.go#L22

Reply:
Should be addressed in 10405bf4296cb7e67e46aea908f61a79778dda29, can you check it please?

Reply:
Uh yeah. I definitely kind of missed that.